src/services/cohortCreation/index.ts

```diff
@@ -1,10 +1,120 @@
+  const patientFilter = `_count=${PATIENT_MAX_COUNT}&_has:Encounter:subject:service-provider=${perimeters.join(',')}`
+  const criteriaGroup = query.request[0].criteria
+
+  const aggregatePatients = async (
+    group: QueryGroup,
+    patientIds: (string | undefined)[],
+    currentCohort: (string | undefined)[]
+  ): Promise<(string | undefined)[]> => {
+    switch (group._type) {
+      case 'andGroup':
+        if (group.isInclusive)
+          return currentCohort.length === 0 ? [...patientIds] : intersection(currentCohort, patientIds)
+        else
+          return currentCohort.length === 0
+            ? [...pullAll(await getPatients(`/Patient?${patientFilter}`), patientIds)]
+            : [...pullAll(currentCohort, patientIds)]
+      case 'orGroup':
+        return group.isInclusive
+          ? union(currentCohort, patientIds)
+          : union(currentCohort, pullAll(await getPatients(`/Patient?${patientFilter}`), patientIds))
+      default:
+        return [...currentCohort]
+    }
+  }
+
+  const queryByResourceType = (criteria: QueryGroup): string => {
+    switch (criteria.resourceType) {
+      case 'Patient':
+        return `/Patient?${patientFilter}&${criteria.filterSolr}`
+      case 'Condition':
+        return `/Patient?${patientFilter}&_has:Condition:patient:${criteria.filterSolr}`
+      default:
+        return ''
+    }
+  }
+
+  const queries = criteriaGroup.reduce<string[]>((acc, group) => {
+    return [...acc, ...group.criteria.reduce<string[]>((acc, criteria) => [...acc, queryByResourceType(criteria)], [])]
+  }, [])
+  const fetchPatients = async (queries: string[]) => await Promise.all(queries.map((query) => getPatients(query)))
+  await fetchPatients(queries)
+
+  const patientIds = await criteriaGroup
+    .sort((a, b) => Number(b.isInclusive) - Number(a.isInclusive))
+    .reduce(async (patientIdsAcc, group) => {
+      const groupPatientIds = await group.criteria
+        .sort((a, b) => Number(b.isInclusive) - Number(a.isInclusive))
+        .reduce(async (groupPatientIdsAcc, criteria) => {
+          const query = queryByResourceType(criteria)
+          const patientIds = await getPatients(query)
+
+          return await aggregatePatients(
+            {
+              ...criteria,
+              _type: group._type
+            },
+            patientIds,
+            await groupPatientIdsAcc
+          )
+        }, Promise.resolve<(string | undefined)[]>([]))
+
+      return await aggregatePatients(
+        {
+          ...group,
+          _type: query.request[0]._type
+        },
+        groupPatientIds,
+        await patientIdsAcc
+      )
+    }, Promise.resolve<(string | undefined)[]>([]))
+
+  return {
+    resourceType: 'Group',
+    type: GroupTypeKind._person,
+    actual: true,
+    quantity: patientIds.length,
+    member: patientIds.map(
+      (id): IGroup_Member => ({
+        entity: {
+          reference: `Patient/${id}`
+        }
+      })
+    )
+  }
```

Comment:
I can put the perimeters in there although I'm not quite sure on the code to use and I'm not comfortable using custom codes. But I won't put every criteria in there since I would need several nested Group resources to do that (the custom json contains multiple nested groups with different relationships between them). If we're going this way, we can ditch the custom json altogether.

Reply:
imo only the perimeters is good enough for now, we'll add nested criteria groups later. If you want, you can also ditch the custom json in there, I'm ok with that